sdk/metric/sdk.go

```diff
@@ -562,23 +563,25 @@ func (m *SDK) collectRecords(ctx context.Context) int {
 
 	m.current.Range(func(key interface{}, value interface{}) bool {
 		inuse := value.(*record)
-		unmapped := inuse.refMapped.tryUnmap()
-		// If able to unmap then remove the record from the current Map.
-		if unmapped {
-			// TODO: Consider leaving the record in the map for one
-			// collection interval? Since creating records is relatively
-			// expensive, this would optimize common cases of ongoing use.
-			m.current.Delete(inuse.mapkey())
-		}
 
-		// Always report the values if a reference to the Record is active,
-		// this is to keep the previous behavior.
-		// TODO: Reconsider this logic.
-		if inuse.refMapped.inUse() || atomic.LoadInt64(&inuse.modified) != 0 {
-			atomic.StoreInt64(&inuse.modified, 0)
+		mods := atomic.LoadInt64(&inuse.updateCount)
+		coll := inuse.collectedCount
+
+		if mods != coll {
+			// Updates happened in this interval.
 			checkpointed += m.checkpointRecord(ctx, inuse)
+		} else {
+			// No updates since last collection.
+			unmapped := inuse.refMapped.tryUnmap()
+
+			// If able to unmap then remove the record from the current Map.
+			if unmapped {
+				m.current.Delete(inuse.mapkey())
+			}
```

Comment:
We should probably jump to the next iteration here, no? Otherwise we do this `inuse.collectedCount = mods` on an unmapped record, which feels like a use-after-free of sorts.

```suggestion
				m.current.Delete(inuse.mapkey())
				return true
			}
```

Reply:
I refactored this function a bit more. There was a bug in the code already, but adding return here is the same: there's a potential for a caller to bind and unbind a handle between loading the atomic value and unreferencing the record here. Now the code re-loads the `updateEpoch` after unreferencing the record (but does not update `collectedCount` "after free").